File: business/src/test/java/com/radioteria/test/business/services/userService/UserActivationTest.java
package com.radioteria.test.business.services.userService;

import com.radioteria.business.events.userService.UserConfirmedEvent;
import com.radioteria.business.services.user.exceptions.UserNotFoundException;
import com.radioteria.db.entities.User;
import com.radioteria.db.enumerations.UserState;
import org.junit.Test;
import org.mockito.ArgumentCaptor;

import java.util.Optional;

import static org.junit.Assert.*;

import static org.mockito.Mockito.*;

public class UserActivationTest extends AbstractUserServiceTest {

    @Test
    public void testActivation() {

        User user = createInactiveUser();

        configureDaoToReturnGivenUser(user);

        userService.activateByEmail("foo@bar.com");

        verifyThatUserIsActivated(user);
        verifyThatEventFiredWithUser(user);

    }

    @Test(expected = UserNotFoundException.class)
    public void testActivationWhenUserDoesNotExist() {

        configureDaoToReturnNull();

        userService.activateByEmail("foo@bar.com");

    }

    private User createInactiveUser() {

        User user = new User("foo@bar.com");
        user.setState(UserState.INACTIVE);

        return user;

    }

    private void configureDaoToReturnGivenUser(User user) {

        when(userDao.findByEmail(anyString())).thenReturn(Optional.of(user));

    }

    private void configureDaoToReturnNull() {

        when(userDao.findByEmail(anyString())).thenReturn(Optional.empty());

    }

    private void verifyThatUserIsActivated(User user) {

        assertEquals(UserState.ACTIVE, user.getState());

    }

    private void verifyThatEventFiredWithUser(User user) {

        ArgumentCaptor<UserConfirmedEvent> eventCaptor = ArgumentCaptor.forClass(UserConfirmedEvent.class);

        verify(eventPublisher, times(1)).publishEvent(eventCaptor.capture());

        UserConfirmedEvent capturedEvent = eventCaptor.getValue();

        assertSame(user, capturedEvent.getUser());

    }


}

File: business/src/test/kotlin/com/radioteria/test/services/user/UserActivationTest.kt
package com.radioteria.test.services.user

import com.radioteria.db.enums.UserState
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.springframework.transaction.annotation.Transactional
import kotlin.test.assertEquals

open class UserActivationTest : BaseUserServiceTest() {

    @Before
    fun before() {
        userService.register("foo@bar.com", "", "")
    }

    @Test
    @Transactional
    open fun activateUser() {

        userService.activateByEmail("foo@bar.com")

        val activatedUser = userService.findByEmail("foo@bar.com")

        if (activatedUser == null) {
            Assert.fail()
            return
        }

        assertEquals(UserState.ACTIVE, activatedUser.state)

    }

}
